### PR TITLE
Enhance Auto-Round

### PR DESCRIPTION
**Pull request by yiliu30**
This PR includes several enhancements for `Auto-round`: 

#### 1. Bring `torch.compile` to speed up the `Auto-round` optimization process: 
```bash
python torchao/prototype/autoround/autoround_llm.py -c
```

- `meta-llama/Llama-2-7b-chat-hf`, about 1.29x ,  27 sec/block -> 21 sec/block
- `meta-llama/Meta-Llama-3.1-8B-Instruct`, about 1.23x, 32 sec/block -> 26 sec/block

https://github.com/yiliu30/torchao-fork/pull/18

#### 2. Add `AO_USE_DETERMINISTIC_ALGORITHMS` for reproducing the `lm-eval` results:  
```bash
AO_USE_DETERMINISTIC_ALGORITHMS=1 python torchao/prototype/autoround/eval_autoround.py
```
https://github.com/yiliu30/torchao-fork/pull/19

#### 3. Expose `gradient_accumulate_steps` to users and update results:
- For `meta-llama/Meta-Llama-3.1-8B-Instruct`, here are the updated results:

|                  | Avg.   | Mmlu   | Piqa   | Winogrande | Hellaswag | Lambada_openai |
| ---------------- | ------ | ------ | ------ | ---------- | --------- | -------------- |
| bf16             | 0.7080 | 0.6783 | 0.8003 | 0.7403     | 0.5910    | 0.7303         |
| torchao-int4wo   | 0.6883 | 0.6363 | 0.7938 | 0.7348     | 0.5784    | 0.6980         |
| autoround-4bit   | 0.6996 | 0.6669 | 0.7916 | 0.7285     | 0.5846    | 0.7262         |
| autoround-4bit*  | 0.7010 | 0.6621 | 0.7976 | 0.7316     | 0.5847    | 0.7291         |

For more details, please refer [README.md](https://github.com/yiliu30/torchao-fork/tree/enhance/autoround/torchao/prototype/autoround#meta-llamameta-llama-31-8b-instruct)

https://github.com/yiliu30/torchao-fork/pull/20

#### TODO: 
- [x] Add `gradient_accumulate_steps` and `compile_optimization_process` to `eval.py` after https://github.com/pytorch/ao/pull/868 lands.  Done by [a5722d8](https://github.com/pytorch/ao/pull/870/commits/a5722d8cf04554a452beb1a00289d78e6388b9f0)

cc @wenhuach21 @thuang6

### CODE REVIEW

**Comment by wenhuach21** [comment]
The reason why accumualte_gradient is better is:   
to better align with the ao's API, we switched from random sampling in our original training implementation to using fixed samples within each batch, which reduces flexibility. However, gradient accumulation helps to partially recover this flexibility.

**Comment by yiliu30** [comment]
Hi @jerryzh168, could you please take a look, thanks!

File: torchao/prototype/autoround/README.md [hunk 1]
```diff
@@ -71,31 +71,36 @@ quantize_(model, apply_auto_round(), is_target_module)
...
+| torchao-int4wo   | 0.6252 | 0.4427 | 0.7617 | 0.6654     | 0.5674    | 0.6889         |
+| autoround-4bit   | 0.6311 | 0.4548 | 0.7606 | 0.6614     | 0.5717    | 0.7072         |
+| autoround-4bit*  | 0.6338 | 0.4566 | 0.7661 | 0.6646     | 0.5688    | 0.7130         |
 
 > [!NOTE]
-> - `auto-round-4bit` represents the following configuration: `bits=4`, `iters=200`, `seqlen=2048`, `train_bs=8`, `group_size=128`, and `quant_lm_head=False`. <br>
 > - `torchao-int4wo` represents `int4_weight_only(group_size=128)` and `quant_lm_head=False`.
-> - If the model includes operations without a deterministic implementation (such as Flash Attention), the results may differ slightly.
+> - `auto-round-4bit` represents the following configuration: `bits=4`, `iters=200`, `seqlen=2048`, `train_bs=8`, `group_size=128`, and `quant_lm_head=False`. <br>
+> - `auto-round-4bit*` represents the following configuration: `bits=4`, `iters=200`, `seqlen=2048`, `train_bs=4`,  `gradient_accumulate_steps=2`, `group_size=128`, and `quant_lm_head=False`. <br>
```
Comment (jerryzh168) [inline torchao/prototype/autoround/README.md:101]: can you just spell out the command for these?

File: torchao/prototype/autoround/core.py [hunk 2]
```diff
@@ -94,7 +98,9 @@ def prepare_model_for_applying_auto_round_(
         group_size (int, optional): The group size for quantization. Defaults to 128.
         iters (int, optional): The number of iterations for optimization. Defaults to 200.
         use_optimized_layer_output (bool, optional): Whether to use optimized layer output. Defaults to False.
-        device (Optional[torch.types.Device], optional): The device to use for accelrating optimization and calibration.
+        gradient_accumulate_steps (Optional[int]): The number of gradient accumulation steps. Defaults to 1.
```
Comment (jerryzh168) [inline torchao/prototype/autoround/core.py:101]: maybe a little bit more docs for the context of gradient accumulation might be helpful, e.g. when this is used in autoround etc.

**Review by jerryzh168 — APPROVED** [review]
LGTM!

File: torchao/prototype/autoround/README.md (hunk 1 above)
Comment (yiliu30) [reply torchao/prototype/autoround/README.md]: Thanks, updated, I also added a table to describe the details of each args. https://github.com/yiliu30/torchao-fork/tree/enhance/autoround/torchao/prototype/autoround#quick-start

File: torchao/prototype/autoround/core.py (hunk 2 above)
Comment (yiliu30) [reply torchao/prototype/autoround/core.py]: Thanks, updated.
